### DISCUSSION TITLE
Implementierung der verschiedenen Kamera-Daten

### DISCUSSION THREAD

**Discussion by KonKob** [Defense-Circuits-Lab/Gait_Analysis#13] [Ideas]
1) Was machen wir mit Körperteilen, die von mehreren Kameras (topcam/sidecam/bottomcam) getreckt werden? 
-  a) Körperteil in allen Kameras: z.B. Snout, Tail
     - Tail am besten sichtbar in TopCam (einheitlicher Background, kein Gegenlicht) -> xy-Koordinaten nur aus TopCam nutzen und mit Sidecam z Koordinaten ergänzen?
     - Wenn Tail Data in TopCam excluded: bottomcam data nutzen?
-  b) Körperteil in Sidecam + nur einer weiteren Kamera: z.B. Pfoten, (Ohren)
    - xy-Koordinaten aus der topcam/bottomcam nutzen und z Koordinaten aus der TopCam (BottomCam: Pfoten, CoG; TopCam: Snout, Tail, Ears)

2) Wie behandeln wir verschiedene Framerates? 
-  a) Derzeit gibt es eine Reihe pro Frame im Dataframe. In einer zusätzlichen Series "Time" wird die Zeit seit t=0 zu den jeweiligen Frames gespeichert. Wie machen wir Berechnungen, die mehrere Kamera-Daten verwenden?

3) Funktionen:
-  a) Richtung (Orientierung der Maus im Maze): 3 Optionen:
        - nur die Richtung der TopCam (bzw. einer einzelnen anderen Kamera) verwenden (derzeit implementiert)
        - Richtung doublechecken, d.h. Richtung muss in Top- und Bottom-Cam übereinstimmen, sonst ungültig
        - wenn Richtungsberechnung in einer Kamera in einem Frame aufgrund von fehlenden (excluded) Werten nicht möglich ist,   Richtungsberechnung aus anderer Kamera für diesen Frame nutzen
-  b) Freezing-Definition: immobility of 
     - EarRight + EarLeft + cCoG (tc), derzeit implementiert
     - EarRight/Left + TailBase (tc) + CoG (bc) + Snout (sc) o.ä.
-  c) GaitDisruption-Definition:
     - ==freezing, derzeit implementiert
     - alle Paws immobile
     - alle Paws immobile + Freezing

**Comment by DennisDoll** [comment]
Ich würde noch ein paar Punkte hinzufügen, die in der Analyse-Pipeline noch vor den oben genannten kommen:

0.1) Zeitliche Synchronisation der Kameras - steht der Plan noch, es mittels eines LED Lämpchens zu machen, das alle Kameras sehen? Sofern ja:
- Wir sollten das Lämpchen mehrfach während des Recordings blinken lassen, zB alle 60 Sekunden. Das würde uns die Möglichkeit geben über das Video hinweg immer wieder zu überprüfen, ob die Frames tatsächlich noch synchron sind (vs. nur 1-2 x Blinken am Anfang, wodurch wir mit unterschiedlichen Framerates womöglich einen Drift erhalten könnten, sodass die Frames am Ende des Recordings nicht mehr so gut zusammen passen

0.2) Räumliche Synchronisierung der Kameras:
- Kein Muss, aber durch eure coole Idee das Setup in sich zu fixieren (bis auf die Top-View Camera haben alles Cams immer die selbe relative Position zueinander [korrket?]) hätten wir die Option das Ganze in Richtung [Stereovision](https://en.wikipedia.org/wiki/Computer_stereo_vision) zu erweitern. Mag einige sehr coole Features mit sich bringen, mag uns ein weiteres Edge für eine Publikation bringen, mag vielleicht aber auch einfach nur zu viel Arbeit/Probleme kreieren ohne entsprechenden Mehrwert zu liefern
- How to? Hab mal ein paar Links rausgesucht:
  - [Paper I](https://lmb.informatik.uni-freiburg.de/projects/student/labcourse/2015Frey.pdf)
  - [Paper II](https://people.inf.ethz.ch/pomarc/pubs/LiIROS13a.pdf)
  - abandoned [StereoVision GitHub Repo](https://github.com/erget/StereoVision)
  - [Tutorials](https://erget.wordpress.com/2014/02/01/calibrating-a-stereo-camera-with-opencv/) + more in the Readme of the Repo
  - The math behind it [TowardsDataScience](https://towardsdatascience.com/camera-calibration-with-example-in-python-5147e945cdeb)
- Alternativ: wir normalisieren für jede Kamera den Maze-Space (ähnlich wie es in dem Code von Michaels Thesis gemacht wird) und hoffen dass die Punkte dann halbwegs gut zueinander passen
- Vermutung: selbst wenn wir den ganzen StereoVision Weg am Ende nicht gehen, wird es vermutlich besser sein alle Bedingungen dafür zunächst zu erfüllen. Dann könnten wir immer noch den alternativen Weg gehen und wenn wir merken, dass das zu ungenau ist, hätten wir die Option auf proper Stereovision zu wechseln. 

0.3) Eig kein wirklicher Diskussionspunkt, denn einzelne Modelle werden wohl unumgänglich sein: Trainieren wir ein DLC Modell für jede Kamera? Oder Eines was auf allen funktioniert? Ein Marker sollte dabei bei allen Modellen auf der Synchronisations-LED liegen, dann haben wir das direkt im DLC output file

**Reply by KonKob** [reply]
Wenn wir wirklich in die Richtung gehen wollen, könnten wir auch gleich überlegen, 3DLabCut (https://deeplabcut.github.io/DeepLabCut/docs/Overviewof3D.html) zu verwenden (mit der Erweiterung Anipose(https://anipose.readthedocs.io/en/latest/index.html)). In dem Fall erstellen wir tatsächlich nur ein (neues) Deeplabcut-Netzwerk, brauchen allerdings einige Kalibrierungs-Videos mit Checkerboards und bekommen als Output einen normalisierten Stereovision 3D-Space. 
Zu zwei Punkten habe ich nichts gefunden: verschiedene Framerates und temporale Synchronisierung, allerdings denke ich, dass basierend auf den kalibrierten Koordinaten sogar die temporale Synchronisierung erreicht wird.

**Comment by DennisDoll** [comment]
Mein Hirn brauchte gerade mal eine Pause von einem anderen Projekt, von daher habe ich mir nochmal ein paar Gedanken zu Deinen (@KonKob) [Punkten](https://github.com/KonKob/Analysis_B06/discussions/13#discussion-4197409) gemacht.

Für Körperteile die von mehreren Cams und dementsprechend unterschiedlichen DLC Modellen getrackt wurden, sollten wir generell auch versuchen alle Infos die wir haben zu nutzen anstatt generell zu sagen die von der einen oder von der anderen Cam schließen wir aus. Denn warum trainieren wir das jeweilige Netz dann überhaupt erst darauf? Stattdessen würde ich vorschlagen, wir versuchen für jeden Marker eine Gewichtung zu erstellen (zwei Vorschläge wie die Gewichtung zustandekommen könnte weiter unten). Wenn wir von einem Marker dann die Koordinaten von mehreren Cams (= DLC Modellen) in einem unified coordinate system haben (entweder Stereovision oder eben jeweils einzeln auf den Maze-Space normalisiert/gemappt) können wir Anhand der Gewichtung der unterschiedlichen Cams und obendrein der predicted probability des jeweiligen DLC Modells für diesen Datenpunkt die wahrscheinlichste Position des Markers im unified Space ermitteln. Angenommen also die Modelle von Cam A und B haben beide den Marker für Bodypart XY zum Zeitpunkt t mit einer Probability von 0.98 erkannt, die jeweiligen gesetzten Koordinaten stimmen aber nicht exakt überein, so könnten wir anhand unserer Gewichtung für den Punkt sagen, dass Cam A normalerweise diesen Punkt besser einschätzt und wir daher nicht den Mittelpunkt zwischen den beiden Schätzungen nehmen, sondern einen Punkt der etwas näher an der Schätzung von Cam A liegt. Gleichermaßen könnte man sagen obwohl die Gewichtung für diesen Marker normalerweise zu Gunsten von Cam A ausfällt, wenn das Modell von Cam A an diesem Datenpunkt allerdings bedeutend schlechter sein sollte als das von Cam B, dann gleicht sich das entsprechend aus und man nimmt eben doch das Mittel oder setzt den Punkt sogar etwas näher an die Schätzung von B. Macht das so Sinn für euch? Klingt für mich recht solide gerade :)

Zwei Otionen, wie wir zu den Gewichtungen kommen könnten: 
1. Wir legen die einfach manuell fest (am schnellsten gemacht - aber natürlich auch extrem fehleranfällig / gebiased)
2. Jeder von uns schaut sich, sagen wir 100 random frames von den verschiedenen Cams samt deren Marker Predictions an (vielleicht nur welche mit DLC likelihoods von 0.95 oder mehr) & wir bewerten für jeden einzelnen Marker ob er korrekt positioniert ist. Am Ende haben wir dann entweder 100 dreifach gelabelte (oder alternativ 300 einzeln gelabelte, sollte beides funktionieren - vielleicht wären hier mehr einzeln gelabelte sogar besser) und können davon einfach die Fehlerrate der einzelnen Cams für den jeweiligen Marker bestimmen und somit empirisch die Gewichtung festlegen (geringste Fehlerrate hat natürlich die höchste Gewichtung).

**Reply by KonKob** [reply]
Ja, in die Richtung hatte ich auch überlegt, als ich 1a) Unterpunkt 2 geschrieben habe. Klingt für mich sehr sinnvoll.
